### PR TITLE
plan: refine the use of idAllocator

### PR DESCRIPTION
**Pull request by hanfei1991**

### CODE REVIEW

File: plan/allocator.go [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package plan
+
+import "github.com/pingcap/tidb/context"
+
+type keyType int
```
Comment (XuHuaiyu) [inline plan/allocator.go]: s/ keyType/ idAllocKeyType

File: plan/allocator.go [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+const key keyType = 0
+
+func (k keyType) String() string {
+	return "plan-allocator"
```
Comment (XuHuaiyu) [inline plan/allocator.go]: operator ID allocator?

File: plan/allocator.go [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+func (a *idAllocator) allocID() int {
+	a.id++
+	return a.id
+}
+
+func resetAllocator(ctx context.Context) {
+	ctx.SetValue(key, new(idAllocator))
+}
+
+func getID(ctx context.Context) int {
```
Comment (XuHuaiyu) [inline plan/allocator.go]: 1. getOperatorID?
2. should we first check whether ctx.Value(key) != nil

**Comment by coocood** [comment]
How about use `SessVars` to store the allocator?
So type assertion can be avoided.

**Comment by XuHuaiyu** [comment]
PTAL @coocood

**Comment by coocood** [comment]
LGTM

**Pull request merged**